### PR TITLE
feat: add upscale feature

### PR DESCRIPTION
**Pull request by MaikoTan**

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -80,4 +80,28 @@ export namespace StableDiffusionWebUI {
     parameters: any
     info: any
   }
+
+  /**
+   * @see https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/master/modules/api/models.py#L122
```
Comment (shigma) [inline src/types.ts:85]: prefer permalink because master code is subject to change

Comment (MaikoTan) [reply src/types.ts:85]: OK, will change later.

File: src/index.ts [hunk 2]
```diff
@@ -351,4 +351,63 @@ export function apply(ctx: Context, config: Config) {
     cmd._options.sampler.fallback = config.sampler
     cmd._options.sampler.type = Object.keys(config.type === 'sd-webui' ? sampler.sd : sampler.nai)
   }, { immediate: true })
+
+  if (config.type === 'sd-webui' && config.enableUpscale) {
```
Comment (MaikoTan) [inline src/index.ts:355]: change to
```
ctx.intersect(() => config.type === 'sd-webui')
  .command('novelai.upscale')
```

File: src/config.ts [hunk 3]
```diff
@@ -195,6 +216,17 @@ export const Config = Schema.intersect([
 
   PromptConfig,
 
+  Schema.union([
+    Schema.object({
+      type: Schema.const('sd-webui'),
+      enableUpscale: Schema.boolean().description('是否启用图片放大功能。').default(false),
```
Comment (MaikoTan) [inline src/config.ts:222]: no enable, default to enable.

File: src/config.ts [hunk 4]
```diff
@@ -67,6 +67,25 @@ export namespace sampler {
   }
 }
 
+export const sdUpscalers = [
```
Comment (MaikoTan) [inline src/config.ts:70]: stick to list

File: src/index.ts [hunk 5]
```diff
@@ -351,4 +351,72 @@ export function apply(ctx: Context, config: Config) {
...
+          method: 'POST',
+          timeout: config.requestTimeout,
+          headers: {
+            ...config.headers,
+          },
+          data,
+        })
+        const base64 = stripDataPrefix((resp.data as StableDiffusionWebUI.ExtraSingleImageResponse).image)
+        return segment.image('base64://' + base64)
+      } catch (e) { console.log (e) }
```
Comment (shigma) [inline src/index.ts:415]: It's better to handle the error, like `return session.text('.unknown-error')`.

Also you should use `logger.warn()` instead of `console.log()`.

File: src/index.ts [hunk 6]
```diff
@@ -351,4 +351,72 @@ export function apply(ctx: Context, config: Config) {
...
+    })
+
+  ctx.accept(['upscaler'], (config) => {
+    subcmd._options.upscaler.fallback = config.upscaler
+    subcmd._options.upscaler.type = sdUpscalers
```
Comment (shigma) [inline src/index.ts:420]:
```suggestion
```

This is not related to config change. You should write them above.

File: src/index.ts [hunk 7]
```diff
@@ -351,4 +351,72 @@ export function apply(ctx: Context, config: Config) {
...
+
+  const subcmd = ctx.intersect(() => config.type === 'sd-webui')
+    .command('novelai.upscale')
+    .shortcut('放大', { fuzzy: true })
+    .option('scale', '-s <scale:number>', { fallback: 2 })
+    .option('resolution', '-r <resolution>', { type: resolution })
+    .option('crop', '-c', { type: 'boolean', fallback: true })
+    .option('upscaler', '-u <upscaler>')
+    .option('upscaler2', '-p <upscaler2>')
+    .option('upscaler2visibility', '-v <upscaler2visibility:number>')
```
Comment (shigma) [inline src/index.ts:363]: I don't think it's good to take up many letters.

**Review by shigma — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 7 above)
Comment (MaikoTan) [reply src/index.ts]: No thought to make a new name, anyone suggests?

File: src/config.ts [hunk 8]
```diff
@@ -197,6 +196,7 @@ export const Config = Schema.intersect([
     Schema.object({
       type: Schema.const('sd-webui'),
       sampler: sampler.createSchema(sampler.sd),
+      upscaler: Schema.union(sdUpscalers.map((up) => Schema.const(up))).description('默认的放大算法。').default('Lanczos'),
```
Comment (shigma) [inline src/config.ts:199]:
```suggestion
      upscaler: Schema.union(upscalers).description('默认的放大算法。').default('Lanczos'),
```

Array of primitives will be cast into const.

Comment (MaikoTan) [reply src/config.ts]: That's good.

**Review by shigma — APPROVED** [review]

File: src/index.ts [hunk 9]
```diff
@@ -351,4 +351,70 @@ export function apply(ctx: Context, config: Config) {
...
+
+  const subcmd = ctx.intersect(() => config.type === 'sd-webui')
+    .command('novelai.upscale')
+    .shortcut('放大', { fuzzy: true })
+    .option('scale', '-s <scale:number>', { fallback: 2 })
+    .option('resolution', '-r <resolution>', { type: resolution })
+    .option('crop', '-c', { type: 'boolean', fallback: true })
+    .option('upscaler', '-u <upscaler>', { type: upscalers })
+    .option('upscaler2', '-p <upscaler2>', { type: upscalers })
+    .option('upscaler2visibility', '-v <upscaler2visibility:number>')
```
Comment (shigma) [inline src/index.ts:363]:
```suggestion
    .option('upscaler', '-1 <upscaler>', { type: upscalers })
    .option('upscaler2', '-2 <upscaler2>', { type: upscalers })
    .option('visibility', '-v <visibility:number>')
```

File: src/index.ts [hunk 10]
```diff
@@ -351,4 +351,70 @@ export function apply(ctx: Context, config: Config) {
...
+        image: image.dataUrl,
+        resize_mode: !!options.resolution ? 1 : 0,
+        show_extras_results: true,
+        upscaling_resize: options.scale,
+        upscaling_resize_h: options.resolution?.height,
+        upscaling_resize_w: options.resolution?.width,
+        upscaling_crop: options.crop,
+        upscaler_1: options.upscaler,
+        upscaler_2: options.upscaler2 ?? 'None',
+        extras_upscaler_2_visibility: options.upscaler2visibility ?? 1,
```
Comment (shigma) [inline src/index.ts:396]:
```suggestion
        extras_upscaler_2_visibility: options.visibility ?? 1,
```

File: src/index.ts [hunk 11]
```diff
@@ -351,4 +351,70 @@ export function apply(ctx: Context, config: Config) {
     cmd._options.sampler.fallback = config.sampler
     cmd._options.sampler.type = Object.keys(config.type === 'sd-webui' ? sampler.sd : sampler.nai)
   }, { immediate: true })
+
+  const subcmd = ctx.intersect(() => config.type === 'sd-webui')
+    .command('novelai.upscale')
+    .shortcut('放大', { fuzzy: true })
+    .option('scale', '-s <scale:number>', { fallback: 2 })
+    .option('resolution', '-r <resolution>', { type: resolution })
+    .option('crop', '-c', { type: 'boolean', fallback: true })
```
Comment (shigma) [inline src/index.ts:360]:
```suggestion
    .option('crop', '-C, --no-crop', { value: false, fallback: true })
```

由于默认启用所以这里应该是反向参数。

File: src/index.ts [hunk 12]
```diff
@@ -351,4 +351,70 @@ export function apply(ctx: Context, config: Config) {
...
+    .option('upscaler', '-u <upscaler>', { type: upscalers })
+    .option('upscaler2', '-p <upscaler2>', { type: upscalers })
+    .option('upscaler2visibility', '-v <upscaler2visibility:number>')
+    .option('upscaleFirst', '-f', { type: 'boolean', fallback: false })
```
Comment (shigma) [inline src/index.ts:364]:
```suggestion
    .option('upscaleFirst', '-f', { fallback: false })
```

不带参数自动 boolean。
